go-controller/pkg/network-controller-manager/network_controller_manager.go

```diff
@@ -70,14 +68,127 @@ type NetworkControllerManager struct {
 	// used for leader election
 	identity string
 
-	// controller for all networks, key is netName of net-attach-def, value is *Controller
-	// this map is updated either at the very beginning of ovnkube-master when initializing the default controller
-	// or when net-attach-def is added/deleted. All these are serialized and no lock protection is needed
-	ovnControllers map[string]NetworkController
+	defaultNetworkController BaseNetworkController
+
+	// net-attach-def controller handle net-attach-def and create/delete network controllers
+	nadController *netAttachDefinitionController
+}
+
+func (cm *networkControllerManager) NewNetworkController(nInfo util.NetInfo,
+	netConfInfo util.NetConfInfo) (NetworkController, error) {
+	cnci := cm.newCommonNetworkControllerInfo()
+	topotype := netConfInfo.GetTopologyType()
+	if netConfInfo.GetTopologyType() == ovntypes.Layer3AttachDefTopoType {
+		return ovn.NewSecondaryLayer3NetworkController(cnci, nInfo, netConfInfo), nil
+	}
+	return nil, fmt.Errorf("topology type %s not supported", topotype)
+}
+
+func (cm *networkControllerManager) NewDummyNetworkController(topoType, netName string) (NetworkController, error) {
+	cnci := cm.newCommonNetworkControllerInfo()
+	netInfo := util.NewNetInfo(&ovncnitypes.NetConf{NetConf: types.NetConf{Name: netName}, IsSecondary: true})
+	switch topoType {
+	case ovntypes.Layer3AttachDefTopoType:
+		return ovn.NewSecondaryLayer3NetworkController(cnci, netInfo, &util.Layer3NetConfInfo{}), nil
+	}
+	return nil, fmt.Errorf("topology type %s not supported", topoType)
+}
+
+// Find all the OVN logical switches/routers for the secondary networks
+func findAllSecondaryNetworkLogicalEntities(nbClient libovsdbclient.Client) ([]*nbdb.LogicalSwitch,
+	[]*nbdb.LogicalRouter, error) {
```

Comment:
Will we ever want to get more logical entity types from OVN ? Or just switches / routers ?

Reply:
possible. then we'd probably need to update this function.

Reply:
Hm... I really dislike multiple return values (other than returning `X, error`). 

This will get really hard to track if it expands; out of the top of my head I can see a new struct with the required data as members. What are your thoughts about that ?

Reply:
we could return a slice of ovsdb objects in the future that satisfy some interface, with one of the required methods on that interface being a Delete or something so then we can generically iterate through each object and just call its delete method. Something we can defer to the future though.